exp/internal/controllers/machinepool_controller_phases.go

```diff
@@ -317,3 +325,255 @@ func (r *MachinePoolReconciler) reconcileInfrastructure(ctx context.Context, clu
+	}
+
+	return nil
+}
+
+// deleteDanglingMachines deletes all MachinePool Machines with an infraRef pointing to an infraMachine that no longer exists.
+// This allows us to clean up Machines when the MachinePool is scaled down.
+func (r *MachinePoolReconciler) deleteDanglingMachines(ctx context.Context, mp *expv1.MachinePool) error {
+	log := ctrl.LoggerFrom(ctx)
+
+	log.V(2).Info("Deleting orphaned machines", "machinePool", mp.Name, "namespace", mp.Namespace)
+	machineList := &clusterv1.MachineList{}
+	labels := map[string]string{
+		expv1.MachinePoolNameLabel:      mp.Name,
+		clusterv1.ClusterNameLabel:      mp.Spec.ClusterName,
+		clusterv1.MachinePoolOwnedLabel: "true",
+	}
+
+	if err := r.Client.List(ctx, machineList, client.InNamespace(mp.Namespace), client.MatchingLabels(labels)); err != nil {
+		return err
+	}
+
+	for i := range machineList.Items {
+		machine := &machineList.Items[i]
+		infraRef := machine.Spec.InfrastructureRef
+
+		// Check this here since external.Get() will complain if the infraRef is empty, i.e. no object kind or apiVersion.
+		if (infraRef == corev1.ObjectReference{}) {
+			// Until the InfraMachinePool populates its selector kind and selector labels, the MachinePool controller can't tell that it supports MachinePool Machines
+			// and will create fallback machines to begin with. When the selector fields gets populated, we want to clean up those fallback machines.
+			log.V(2).Info("Machine has an empty infraRef, this a fallback machine, will delete", "machine", machine.Name, "namespace", machine.Namespace)
+
+			return r.Client.Delete(ctx, machine)
+		}
+
+		_, err := external.Get(ctx, r.Client, &infraRef, mp.Namespace)
+		if err != nil {
+			if !apierrors.IsNotFound(err) {
+				log.Error(err, "unexpected error while trying to get infra machine, will delete anyway", "machine", machine.Name, "namespace", machine.Namespace)
+			}
+
+			log.V(2).Info("Deleting orphaned machine", "machine", machine.Name, "namespace", machine.Namespace)
+			if err := r.Client.Delete(ctx, machine); err != nil {
+				return err
+			}
+		} else {
+			log.V(2).Info("Machine is not orphaned, nothing to do", "machine", machine.Name, "namespace", machine.Namespace)
+		}
+	}
+
+	return nil
+}
+
+// createMachinesIfNotExists creates a MachinePool Machine for each infraMachine if it doesn't already exist and sets the owner reference and infraRef.
+func (r *MachinePoolReconciler) createMachinesIfNotExists(ctx context.Context, mp *expv1.MachinePool, infraMachines []unstructured.Unstructured) error {
+	machineList := &clusterv1.MachineList{}
+	labels := map[string]string{
+		expv1.MachinePoolNameLabel:      mp.Name,
+		clusterv1.ClusterNameLabel:      mp.Spec.ClusterName,
+		clusterv1.MachinePoolOwnedLabel: "true",
+	}
+
+	if err := r.Client.List(ctx, machineList, client.InNamespace(mp.Namespace), client.MatchingLabels(labels)); err != nil {
```

Comment:
My intention was to query twice so that we always have an up to date copy in case the list changed in between querying and our operations. Do you think we would still need to worry about that or could we just use one query?

Reply:
Querying two times doesn't ensure you will have up-to-date data due to how the cache works.
Also, if some change happens in the meantime a new reconciliation will be triggered (assuming watches are properly defined), so reading twice in the middle of a reconciliation IMO could created more issues than benefits